File: docker/ui/app/src/types/stir.ts
/**
 * STIR/SHAKEN attestation types — map 1:1 to the FastAPI response schemas.
 *
 * The per-call "story" is a chain:
 *   caller `inbound_attest` (verified per `inbound_verstat`, sourced from
 *   `verstat_source`)  →  we emitted `signed_attestation`.
 *
 * Example: "Caller A ✓ (carrier-verified) → Signed div".
 */

/** SHAKEN attestation levels. `div` = diversion (call was forwarded/redirected). */
export type AttestationLevel = 'A' | 'B' | 'C' | 'div';

/** Where the inbound verification verdict (`inbound_verstat`) came from. */
export type VerstatSource = 'carrier' | 'self';

/**
 * Per-call attestation — `GET /v1/cdrs/{call_id}/attestation`.
 * The endpoint is tenant-scoped server-side and returns 404 when a call has no
 * attestation record (older calls, unsigned calls, purely on-net calls).
 */
export interface CallAttestation {
  call_id: string;
  customer_id: number;
  /** The attestation level we (the platform) signed the outbound leg with. */
  signed_attestation: AttestationLevel | null;
  /** The attestation level we intended to sign with (before any downgrade). */
  attest_intent: AttestationLevel | null;
  /** Whether the inbound leg arrived with a valid SHAKEN Identity header. */
  inbound_signed: boolean;
  /** The caller's attestation level, as received on the inbound leg. */
  inbound_attest: AttestationLevel | null;
  /**
   * The verification result for the inbound leg, e.g.
   * "TN-Validation-Passed" / "TN-Validation-Failed" / "No-TN-Validation".
   */
  inbound_verstat: string | null;
  /** Origin of the `inbound_verstat` verdict: carrier-supplied or self-verified. */
  verstat_source: VerstatSource | null;
  created_at: string;
}

/** One row of a summary breakdown: a dimension value and how many calls had it. */
export interface AttestationBreakdownItem {
  /** `null` means the dimension was absent for those rows. */
  value: string | null;
  count: number;
}

/**
 * Aggregate attestation summary — `GET /v1/stir/attestation-summary` (admin).
 * Defaults to the last 7 days when no date range is supplied.
 */
export interface AttestationSummary {
  total: number;
  customer_id: number | null;
  start_date: string;
  end_date: string;
  by_signed_attestation: AttestationBreakdownItem[];
  by_inbound_attest: AttestationBreakdownItem[];
  by_inbound_verstat: AttestationBreakdownItem[];
  by_verstat_source: AttestationBreakdownItem[];
}

/** Query params for the admin summary endpoint. All optional. */
export interface AttestationSummaryParams {
  customer_id?: number;
  /** ISO date (YYYY-MM-DD) or full ISO timestamp — the API accepts either. */
  start_date?: string;
  end_date?: string;
}

File: docker/api/src/routers/stir.py
"""STIR/SHAKEN attestation summary endpoints (admin only).

Aggregate visibility over the `call_attestations` companion table
(see docker/postgres/init/32_call_attestations.sql). Per-call reads live on
the CDRs router (GET /v1/cdrs/{call_id}/attestation, tenant-scoped); this
router is the admin roll-up for the UI summary panel.
"""
from datetime import datetime, timedelta, timezone
from typing import Optional

from fastapi import APIRouter, Depends, Query

from auth.dependencies import require_admin
from db import database as db

router = APIRouter()


@router.get("/attestation-summary")
async def attestation_summary(
    customer_id: Optional[int] = None,
    start_date: Optional[datetime] = None,
    end_date: Optional[datetime] = None,
    admin: dict = Depends(require_admin),
):
    """STIR/SHAKEN attestation roll-up (admin only).

    Returns GROUP BY counts over the `call_attestations` table for the given
    window (default: last 7 days), optionally scoped to one customer:
      - by_signed_attestation: what WE emitted (A / B / C / div)
      - by_inbound_attest:     the caller's attestation from the origin carrier
      - by_inbound_verstat:    inbound caller verification result
      - by_verstat_source:     'self' (our crypto) vs 'carrier' (their PAI)

    Each breakdown is a list of {value, count} objects (value=null groups the
    rows where that dimension was not present). Also returns `total` (rows in
    window) and the effective window echoed back.

    Filtering is by created_at (ingest time). Backed by the
    (customer_id, created_at) index for an efficient range scan.
    """
    if not start_date:
        start_date = datetime.now(timezone.utc) - timedelta(days=7)
    if not end_date:
        end_date = datetime.now(timezone.utc)

    # Shared WHERE clause: created_at=$1..$2, optional customer_id=$3.
    where = "WHERE created_at >= $1 AND created_at <= $2"
    values: list = [start_date, end_date]
    if customer_id is not None:
        where += " AND customer_id = $3"
        values.append(customer_id)

    async def _grouped(column: str) -> list[dict]:
        """Return [{value, count}] grouped by one attestation dimension."""
        rows = await db.fetch_all(
            f"""
            SELECT {column} AS value, COUNT(*) AS count
            FROM call_attestations
            {where}
            GROUP BY {column}
            ORDER BY count DESC
            """,
            *values,
        )
        return [{"value": r["value"], "count": r["count"]} for r in rows]

    total_row = await db.fetch_one(
        f"SELECT COUNT(*) AS total FROM call_attestations {where}",
        *values,
    )

    return {
        "total": (total_row["total"] if total_row else 0),
        "customer_id": customer_id,
        "start_date": start_date.isoformat(),
        "end_date": end_date.isoformat(),
        "by_signed_attestation": await _grouped("signed_attestation"),
        "by_inbound_attest": await _grouped("inbound_attest"),
        "by_inbound_verstat": await _grouped("inbound_verstat"),
        "by_verstat_source": await _grouped("verstat_source"),
    }
